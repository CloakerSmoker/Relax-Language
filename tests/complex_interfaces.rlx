#Require "$/Console.rlx"

class AllocateMePlease derives IAllocate {
    i8 Value
}

class Complicated derives INew {
    i32 First
    i64 Second

    define void New() {
        Print("Complicated.New()")

        this->First := 10
        this->Second := 20
    }

    define void New(i32 First) {
        Print("Complicated.New(%i)", First)

        this->First := First
        this->Second := 40
    }

    define void New(i32 First, i64 Second) {
        Print("Complicated.New(%i, %i)", First, Second)

        this->First := First
        this->Second := Second
    }
}

define i32 Main() {
    Print("IAllocate:\n")

    AllocateMePlease* Please := AllocateMePlease:Allocate()

    Please->Value := 20
    Print("Value: %i\n", Please->Value)

    Print("\nINew:\n")

    Print("Zero   Parameter New: ")
    Complicated* C1 := Complicated:New()
    Print("      : %i %i\n", C1->First, C1->Second)

    Print("Single Parameter New: ")
    Complicated* C2 := Complicated:New(30)
    Print("    : %i %i\n", C2->First, C2->Second)

    Print("Double Parameter New: ")
    Complicated* C3 := Complicated:New(50, 60)
    Print(": %i %i\n", C3->First, C3->Second)

    return 0
}