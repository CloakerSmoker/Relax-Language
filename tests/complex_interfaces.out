IAllocate:
Value: 20

INew:
Zero   Parameter New: Complicated.New()      : 10 20
Single Parameter New: Complicated.New(30)    : 30 40
Double Parameter New: Complicated.New(50, 60): 50 60

Default Method Implementation:
B ->GetFavoriteNumber(): IDefineDefaults.GetFavoriteNumber(): 70
BO->GetFavoriteNumber(): IDefineDefaults.GetFavoriteNumber(): 70
O ->GetFavoriteNumber(): Opinionated.GetFavoriteNumber(): 80
OD->GetFavoriteNumber(): Opinionated.GetFavoriteNumber(): 80

TypeID:
B->TypeID()    : 0x401FC6B3A30D0D4
typeid:(Boring): 0x401FC6B3A30D0D4
BD->TypeID()   : 0x401FC6B3A30D0D4
O->TypeID()         : 0xC43503884CB69701
typeid:(Opinionated): 0xC43503884CB69701
OD->TypeID()        : 0xC43503884CB69701

New:
new Complicated()          : Complicated.New()
new Complicated(i32)       : Complicated.New(80)
new Complicated(i32, i32)  : Complicated.New(90, 100)
new Opinionated()->TypeID(): 0xC43503884CB69701

Default Methods Calling Methods
E->GetNumber(): ICMFD.GN(): ICMFD.IGN: 200
